xcbuild: add wrapper

matthewbauer opened the pull request:
###### Motivation for this change

This is going to turn into a pull request for adding "xcbuild" which is an open source Xcode-compatible build tool. Hopefully, we can trick it into accepting our "stdenv" instead of the Mac SDK.
###### Things done
- Built on platform(s)
  - [x] NixOS
  - [x] OS X
  - [x] Linux
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
## TODO
- [x] remove impurities
- [x] get some package to build with xcbuild
- [x] document xcbuild wrapper functionality and usage
## Moving packages to "xcbuild"
- [ ] pinentry-mac (needs storyboard tool)
- [ ] retroarch (needs storyboard tool)
- [ ] macvim (needs perl 5.16)
- [x] contacts
- [ ] chuck
- [ ] xcodeenv?
## New packages
- [x] adv_cmds (not pkill)
- [x] basic_cmds
- [ ] bootstrap_cmds
- [x] developer_cmds
- [ ] diskdev_cmds
- [ ] doc_cmds
- [ ] file_cmds
- [x] network_cmds (not spray, rtadvd)
- [ ] mail_cmds
- [ ] misc_cmds
- [ ] system_cmds

matthewbauer committed:
xcbuild: add wrapper

Also updates xcbuild version.

This changes the raw string expressions into nix expressions that are
then converted into json by builtins.toJSON. Then, converted to Plist
XML by Apple's plutil. Sadly, xcbuild does not support using raw JSON
but Apple's plutil does so we just convert the file from JSON to XML
using Apple's plutil. The result is not ideal but it looks like all OS X
systems have working plutil's.

- set mac version to 10.10
- add setup hook.

copumpkin commented:
Very excited to see where this goes. I vaguely remember last time I looked at `xcbuild` that it wasn't able to build the projects I threw at it because it relied on someone going into xcode and adding some sort of project aspect that was often missing in the xcode projects on opensource.apple.com, but things might have changed in the past couple of years since I first looked.

👍

copumpkin commented:
Oh sorry, I completely forgot about this! Your stuff looks far deeper than mine. I'd be happy to take a closer look at this now that I remember it's there. Or how much of it is still relevant?

copumpkin commented:
Your platform plist stuff was exactly the sort of thing I was planning to look into, so this looks great. I'd probably not make a giant literal string (since we have plutil we can generate it on the fly from json which we can generate from pure nix). Very cool stuff! Come on IRC sometime 😄

matthewbauer commented:
@copumpkin
Getting rid of the raw strings would be really helpful. Is plutil okay to use? Right now I don't think it's available in nixpkgs.

copumpkin commented:
@matthewbauer I just noticed it in the `xcutil` package yesterday when I was poking around. It appears to be a full reimplementation of the original, but I haven't tested it. The command-line options look the same though!

So basically, I'd probably use the `xcutil` package, then use its `plutil` to generate the helper platform files, and then wrap those all up into a nix-level builder function. Or your setup hook is probably fine too come to think of it 👍

matthewbauer commented:
Oh never mind - I didn't realize a plutil was included in xcbuild.

copumpkin commented:
I'd probably even make a dedicated builder to wrap `plutil`, so you can feed it a dictionary somehow and get a plist on disk out of it. The possibilities are endless! Thanks for this work

matthewbauer commented:
Yeah xcbuild looks really promising.

One thing I just found though is this:

```
~/Projects/nixpkgs $ /nix/store/*xcbuild*/bin/plutil -help
usage: plutil -<command> [options] <files>

commands:
  -lint (default)
  -help (this message)
  -p
  -convert <format>
  -insert <key> <value>
  -replace <key> <value>
  -remove <key>
  -extract <key> <format>

values:
  -bool <YES|NO>
  -integer <number>
  -float <number>
  -string <utf8>
  -data <base64>
  -date <iso8601>
  -xml <plist>
  -json <json> (not yet implemented)

formats:
  xml1
  binary1
  openstep1
  json (not yet implemented)
```

It looks like json isn't implemented yet.

copumpkin commented:
Oh, boo! I guess maybe leave your things as strings for now and put a TODO next to them saying to replace them with Nix literals once `plutil` grows JSON support? Or one of us could make a PR upstream adding the JSON support. But it's not really the most pressing thing so honestly just do what you think is best 😄 either way really cool that you generate your own plists for our stuff.

copumpkin commented:
There's also another candidate for `plutil` in https://github.com/apple/swift-corelibs-foundation/tree/master/Tools/plutil but I don't think that's ready yet either.

matthewbauer commented:
@copumpkin The nix expression stuff looks to be working. The problem is I've yet to find any Xcode projects that don't need the ibtool to work correctly. I might try just making a bash script to hook into for ibtool.

copumpkin commented:
@matthewbauer all the opensource.apple.com ones 😄😄😄😄😄

copumpkin commented:
Maybe try something like adv_cmds? I think we have some hackery in there to just build `ps` from it right now, but it would be amazing if you could build that using xcbuild without hackery.

matthewbauer commented:
Another thing I've been having issues with:

DERIVED_DATA_DIR needs to be set to the nix build directory. How do I set that up in wrapProgram so it resolves when xcodebuild is called instead of when the wrapper is created?

copumpkin commented:
@matthewbauer can't you just set it up to be relative to the cwd?

copumpkin commented:
Or maybe just a custom builder that sets things up properly?

matthewbauer commented:
Okay that works- I was just thinking there might be some fancy string escaping that I could use.

matthewbauer commented:
An issue with adv_cmds is it is needed by cmake and cmake is needed by xcbuild. I'm not sure how to resolve that.

copumpkin commented:
Oh, I just meant setting up a new expression for adv_cmds independent of the one in the stdenv. If you can get it working, we can figure out the bootstrap implications afterwards.

matthewbauer committed:
adv_cmds: add xcode derivation

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -564,12 +564,12 @@ stdenv.mkDerivation {
 
-    /usr/bin/plutil -convert xml1 -o Info.plist ${writeText "Info.plist" (builtins.toJSON Info)}
-    /usr/bin/plutil -convert xml1 -o version.plist ${writeText "version.plist" (builtins.toJSON Version)}
+    plutil -convert xml1 -o Info.plist ${writeText "Info.plist" (builtins.toJSON Info)}
```
How does this work? I thought the xcbuild one didn't support JSON yet?

copumpkin commented on pkgs/development/tools/xcbuild/toolchain.nix:
```diff
@@ -58,28 +59,30 @@ stdenv.mkDerivation {
 #   asa
-#   bison
 #   c89
 #   c99
```
I think `c89` and `c99` are entry points for `clang`, but I don't know what produces them.

copumpkin commented on pkgs/development/tools/xcbuild/toolchain.nix:
```diff
@@ -58,28 +59,30 @@ stdenv.mkDerivation {
 #   lipo
 #   lorder
-#   m4
 #   mig
```
This comes from `bootstrap_cmds`

copumpkin commented on pkgs/development/tools/xcbuild/toolchain.nix:
```diff
@@ -93,4 +96,3 @@ stdenv.mkDerivation {
 #   unifdef
```
We have an `unifdef` package

copumpkin commented on pkgs/os-specific/darwin/apple-source-releases/adv_cmds/xcode.nix:
```diff
@@ -0,0 +1,21 @@
+{stdenv, fetchurl, xcbuild}:
+
+stdenv.mkDerivation {
```
Does this expression actually build? Amazeballs if so, and it'll make our lives so much easier.

matthewbauer replied:
No this is not working right now.

copumpkin replied:
Either way, very exciting to see the direction it's going 😄 thanks for doing this!

copumpkin commented on pkgs/development/tools/xcbuild/developer.nix:
```diff
@@ -0,0 +1,134 @@
+    Name = "Lex";
+    Class = "PBXCompilerSpecificationLex";
+    Description = "Lex Scanner Generator";
+    ExecPath = "echo ${flex}/bin/flex";
```
That `echo` seems undesirable 😄

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,578 @@
+            EXECUTABLE_PREFIX = "";
+            EXECUTABLE_SUFFIX = "";
+            REZ_EXECUTABLE = "YES";
+            INSTALL_PATH = "/usr/local/bin";
```
That can't be right!

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,578 @@
+            REZ_EXECUTABLE = "YES";
+            EXECUTABLE_SUFFIX = ".$(EXECUTABLE_EXTENSION)";
+            EXECUTABLE_EXTENSION = "o";
+            PUBLIC_HEADERS_FOLDER_PATH = "/usr/local/include";
```
Nuh uh

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,578 @@
+            EXECUTABLE_EXTENSION = "a";
+            PUBLIC_HEADERS_FOLDER_PATH = "/usr/local/include";
+            PRIVATE_HEADERS_FOLDER_PATH = "/usr/local/include";
+            INSTALL_PATH = "/usr/local/lib";
```
I think you get the point 😄

copumpkin commented on pkgs/development/tools/xcbuild/sdk.nix:
```diff
@@ -0,0 +1,31 @@
+    DisplayName = sdkName;
+    Toolchains = [ toolchainName ];
+    Version = "10.9";
+    MaximumDeploymentTarget = "10.9";
```
I'd aim for 10.10, since that's the version we're using elsewhere. Or if you can pull it in from `MACOSX_DEPLOYMENT_TARGET`, that'd be handy too. Or if not, we could designate a central file in stdenv/darwin/ and perhaps call it target-version.nix, so we can only change it in one place and have it update everywhere.

copumpkin commented:
If you feel like coming onto IRC to discuss this (and anything else about nix on darwin) at some point, we're all friendly in ##nix-darwin on freenode 😄

matthewbauer commented:
@copumpkin Okay I'll try to be on. I didn't realize there was a darwin-specific channel.

matthewbauer commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,578 @@
+            IsLaunchable = "YES";
+        };
+    }
+    {   Type = "PackageType";
```
Yeah, so all of these aren't included right now in xcbuild. But it might be added later.

matthewbauer committed:
xcbuild: Make "xcbuild.raw" for original unwrapped

grp commented on pkgs/development/tools/xcbuild/developer.nix:
```diff
@@ -0,0 +1,142 @@
+    cp ${xcbuild}/Library/Xcode/Specifications/* $out/Contents/Developer/Library/Xcode/Specifications/
+
+    /usr/bin/plutil -convert xml1 ${writeText "Lex.xcspec" (builtins.toJSON Lex)} -o $out/Contents/Developer/Library/Xcode/Specifications/Lex.xcspec
+    /usr/bin/plutil -convert xml1 ${writeText "Yacc.xcspec" (builtins.toJSON Yacc)} -o $out/Contents/Developer/Library/Xcode/Specifications/Yacc.xcspec
```
Did the JSON support from facebook/xcbuild#170 work for this, or anything else you need?

grp replied:
OK, facebook/xcbuild#183 should make this work.

grp commented on pkgs/development/tools/xcbuild/developer.nix:
```diff
@@ -0,0 +1,142 @@
+  ## TAPI
+  ## TiffUtil
+  ## UnifDef
+in
```
For these: I think `LSRegisterURL`, `Ld`, `Libtool`, `Lipo`, and `PBXCp` are implemented already. Some of the others might be more complicated — they aren't used even when xcbuild is run against Xcode's specifications.

grp reviewed (commented):
Just saw this — wow, this is really cool! Definitely let me know if there's anything I can help explain or track down on the xcbuild side.

matthewbauer commented on pkgs/development/tools/xcbuild/developer.nix:
```diff
@@ -0,0 +1,142 @@
+  buildInputs = [ xcbuild ];
+  buildCommand = ''
+    mkdir -p $out/Contents/Developer/Library/Xcode/Specifications/
+    cp ${xcbuild}/Library/Xcode/Specifications/* $out/Contents/Developer/Library/Xcode/Specifications/
```
@grp Is this okay to do? I can't tell if these are actually okay to use for building things or if they are just for xcbuild testing.

grp replied:
They aren't super well tested yet since the set isn't complete, but the hope is that they'd be for exactly this. (If you use xcbuild's CMake install, it can setup the directory structure for you, but not sure if that's common to use with Nix.)

copumpkin replied:
@grp that `${xcbuild}` thing is our build of `xcbuild`, so the directory structure was already set up on it by your cmake scripts

matthewbauer commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,604 @@
+    ProjectName = "OSXPlatformSupport";
+  };
+
+  PackageTypes = [
```
@grp Should these rules be added to xcbuild? It seems like every xcode project will need some of these.

grp replied:
Hard to say — I think they're more part of the `.platform` than the build system that xcbuild provides. But there might be a way to have them as like templates to build off of?

matthewbauer committed:
xcbuild: Update toolchain

- Rework spec files.
- Add more tool specs.
- Add ctags.

copumpkin commented:
Whoa, @grp is commenting in our Nix repository! What has the world come to??

copumpkin commented:
@matthewbauer why include the giant spec files in this repo? I'd much rather fetch them from upstream via a fixed-output derivation than include a bunch of stuff nobody understands that can get out of sync easily 😦

copumpkin commented on pkgs/development/tools/xcbuild/toolchain.nix:
```diff
@@ -0,0 +1,109 @@
+# other commands in /bin/
+#   asa
+#   cmpdylib (in cctools)
+#   ctags (in developer_cmds)
```
We should have an Nixified ctags elsewhere in nixpkgs, right? I wouldn't bother pulling in all of developer_cmds for ctags when it's just a fork of the real one.

matthewbauer commented:
Sorry these are all from different Xcode config sources. They won't be in the final draft I'm just trying to get something that works then slowly remove the unused stuff.

copumpkin commented on pkgs/development/tools/xcbuild/specs/Unifdef.xcspec:
```diff
@@ -0,0 +1,32 @@
+    Unifdef.xcspec
+
+    Copyright (c) 2014 Apple, Inc.
+    All rights reserved.
```
Not sure people will be very happy with an "All rights reserved" Apple file living in this repo 😦 

I'm guessing it doesn't exist anywhere else public? If so, I think our best bet is to shove all these files into a git repo somewhere that we fetch from the xcbuild expression.

copumpkin commented:
@matthewbauer oh, cool. I can leave you alone until you call me back to take a look 😄 just quite excited by this PR as you might have gathered!

matthewbauer commented:
@copumpkin Have you seen this header error at all?

```
In file included from /private/var/folders/l4/v2k_52x55r3_bf4cty3g4bg00000gn/T/nix-build-adv_cmds.drv-14/adv_cmds-163/pkill/pkill.c:68:
/nix/store/rga593w2xn76psfmqpr9xjqrzdk7r1x4-Libsystem-osx-10.9.5/include/os/assumes.h:36:10: fatal error: '_simple.h' file not found
#include <_simple.h>
         ^
1 error generated.
```

It looks like we need the _simple.h header file?

grp commented:
There's an implementation of `os/base_private.h` here: https://github.com/samdmarshall/OSXPrivateSDK/blob/master/PrivateSDK10.10.sparse.sdk/usr/local/include/os/base_private.h

matthewbauer committed:
adv_cmds: Disable pkill.

pkill isn't building because of some missing headers:

- xpc/xpc.h
- os/base_private.h
- _simple.h

They are all available somewhere but not set up correctly in the Darwin
stdenv.

TODO: add pkill back in!

matthewbauer committed:
xcbuild: temporarily fix cflags/ldflags

This gets everything to build.

adv_cmds: remove NIX_LDFLAGS

- unneeded

matthewbauer commented:
@grp Awesome! I can't believe I missed that. We are already packaging that git repo anyway.

But right now, I'm just going to try to disable pkill so we can get something working without modifying the Darwin stdenv.

The builder gets pretty far now but still fails on locale.cc:

```
/nix/store/4lvhalpyvcgl5g12w6ymq8wdggdp64nm-nixpkgs.xctoolchain/usr/bin/clang -x c++ -arch x86_64 -isysroot /nix/store/n9dbwi7vq971gn32jznm2pabafabr9zc-nix.nixpkgs.sdk -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -fasm-blocks -fstrict-aliasing -fpascal-strings -Os -g -Wno-trigraphs -Wno-empty-body -Winfinite-recursion -Wdeprecated-declarations -Wno-missing-field-initializers -Wnewline-eof -Wno-missing-prototypes -Wswitch -Wno-four-char-constants -Wno-conversion -Wno-bool-conversion -Wno-constant-conversion -Wno-enum-conversion -Wno-int-conversion -Wno-sign-conversion -Wno-shorten-64-to-32 -Wno-missing-braces -Wparentheses -Wno-shadow -Wuninitialized -Wunknown-pragmas -Wunused-function -Wunused-label -Wno-unused-parameter -Wunused-value -Wunused-variable -Wno-c++11-extensions -Wno-exit-time-destructors -Winvalid-offsetof -Wno-overloaded-virtual -Wno-non-virtual-dtor -Wno-move -I./adv_cmds-brxplvifngslsmbcffcorvztaldp/Build/Products/Release-/include -I./adv_cmds-brxplvifngslsmbcffcorvztaldp/Build/Intermediates/adv_cmds.build/Release-/locale.build/DerivedSources/x86_64 -I./adv_cmds-brxplvifngslsmbcffcorvztaldp/Build/Intermediates/adv_cmds.build/Release-/locale.build/DerivedSources -F./adv_cmds-brxplvifngslsmbcffcorvztaldp/Build/Products/Release- -MMD -MT dependencies -MF ./adv_cmds-brxplvifngslsmbcffcorvztaldp/Build/Intermediates/adv_cmds.build/Release-/locale.build/Objects-normal/x86_64/locale.d -c /private/var/folders/l4/v2k_52x55r3_bf4cty3g4bg00000gn/T/nix-build-adv_cmds.drv-15/adv_cmds-163/locale/locale.cc -o ./adv_cmds-brxplvifngslsmbcffcorvztaldp/Build/Intermediates/adv_cmds.build/Release-/locale.build/Objects-normal/x86_64/locale.o
/private/var/folders/l4/v2k_52x55r3_bf4cty3g4bg00000gn/T/nix-build-adv_cmds.drv-15/adv_cmds-163/locale/locale.cc:1:10: fatal error: 'iostream' file not found
#include <iostream>
         ^
1 error generated.
```

@copumpkin this seems very similar to #18729. I can't figure out anyway past it though.  This is the first c++ file so that is probably why it is choking there.

grp commented:
Xcode and xcbuild always use the `clang` driver to compile C++, not `clang++`, so I don't think that's the issue. (They do use `clang++` to link.)

Looking at your code, I don't see libc++ going into the toolchain? Clang is going to look for it at a relative path of `../include/c++/v1` relative to its executable: https://github.com/llvm-mirror/clang/blob/3353f9f983c8bf75af288b1ab60e0fdffb1d003a/lib/Frontend/InitHeaderSearch.cpp#L457

matthewbauer committed:
xcbuild: Add install phase

This phase is disabled for now but we can use it later.

matthewbauer committed:
adv_cmds: add custom install phase

matthewbauer commented:
@grp Okays thanks! I think I've got something figured out.

With the exception of pkill all of the commands are building. I still want @copumpkin to take a look at this though because I know that the NIX_CFLAGS_COMPILE isn't a very good way to solve the upstream issue.

For this to really work 100%, we'll want to use the xcodebuild install phase (facebook/xcbuild#62). For now I've just got some copy commands hard coded.

matthewbauer committed:
contacts: use xcbuild instead of xcodebuild

matthewbauer committed:
basic_cmds: init at 55

matthewbauer committed:
adv_cmds: move from all-packages

Use the old way with pkgs/os-specific/darwin/apple-source-releases/defuault.nix.

matthewbauer committed:
Librpcsvc: init at 26

matthewbauer committed:
developer_cmds: init at 62

matthewbauer committed:
network_cmds: init at 481.20.1

Fixes #16764

grp commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,30 @@
+  Info = {
+    CFBundleIdentifier = platformName;
+    Type = "Platform";
+  };
```
It's probably a good idea to set a `Name = "macosx";` or something here. Some code in xcbuild may rely on that being non-empty, IIRC.

(I noticed because a few of the other commits are referencing a directory `Release-`, which is special cased to just `Release` when the platform name is `macosx`.)

matthewbauer commented on pkgs/development/tools/xcbuild/default.nix:
```diff
@@ -16,13 +16,13 @@ let
 in stdenv.mkDerivation rec {
   name    = "xcbuild-${stdenv.lib.substring 0 8 version}";
-  version = "49f8a5923f1381f87ac03ad4c1b138d1d2b74369";
+  version = "ef95e55f4732e26c83b4a8cb266f21c6897b309c";
```
This points to a pr so need to update it once pr is merged.

matthewbauer commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,604 @@
+            DOCUMENTATION_FOLDER_PATH = "$(LOCALIZED_RESOURCES_FOLDER_PATH)/Documentation";
+            PLUGINS_FOLDER_PATH = "$(CONTENTS_FOLDER_PATH)/PlugIns";
+            SCRIPTS_FOLDER_PATH = "$(UNLOCALIZED_RESOURCES_FOLDER_PATH)/Scripts";
+            JAVA_FOLDER_PATH = "$(UNLOCALIZED_RESOURCES_FOLDER_PATH)/Java";
```
Some of these might not be necessary?

matthewbauer commented on pkgs/development/tools/xcbuild/toolchain.nix:
```diff
@@ -0,0 +1,105 @@
+#   lorder
+#   rebase
+#   rpcgen (in developer_cmds)
+#   what
```
rpcgen, what are now packaged in developer_cmds. The trick now is letting these be here and also letting developer_cmds depend on xcbuild.

copumpkin replied:
@matthewbauer the way I see it, once we merge this, we can work on getting a parallel set of "well-built" Apple bootstrap packages using this. Then I (or we) can redo the darwin bootstrap from scratch and figure out how to make this all tie together. I don't expect it'll be too painful, and I'll document the process as we go along. I'm really hoping to automate the generation of bootstrap chains at some point but I'll probably do one more by hand before doing that.

matthewbauer commented on pkgs/os-specific/darwin/apple-source-releases/Librpcsvc/default.nix:
```diff
@@ -0,0 +1,25 @@
+    maintainers = with maintainers; [ matthewbauer ];
+    platforms   = platforms.darwin;
+    license     = licenses.apsl20;
+  };
```
@copumpkin Is it okay if I live out the description meta? some os-specific/darwin/ have it others don't.

copumpkin replied:
Yeah, definitely. I'd just remove the attribute rather than set it to an empty string though.

matthewbauer commented on pkgs/os-specific/darwin/apple-source-releases/adv_cmds/xcode.nix:
```diff
@@ -0,0 +1,37 @@
+  # disable pkill from build
+  patchPhase = ''
+    substituteInPlace adv_cmds.xcodeproj/project.pbxproj \
+      --replace "FD201DC214369B4200906237 /* pkill.c in Sources */," ""
```
pkill may actually be buildable with the extra header files we've found. These are needed:

- xpc/xpc.h
- os/base_private.h
- _simple.h

copumpkin replied:
I'm not too concerned about `pkill`, honestly. For now I'd just leave it the way you have it, possibly with some breadcrumbs in a comment telling people where to find the missing headers. But tbh, since `pkill` isn't needed for bootstrap I probably wouldn't pull in the osx private SDK to the bootstrap just to support `pkill`. Perhaps eventually we can put a conditional dependency on the developer SDK (which I still need to update) and do this more cleanly with official Apple headers, and then only enable `pkill` if the condition is true.

matthewbauer committed:
xcbuild: add name to platform.nix

matthewbauer commented on pkgs/os-specific/darwin/apple-source-releases/adv_cmds/xcode.nix:
```diff
@@ -0,0 +1,37 @@
+  buildInputs = [ xcbuild libcxx ];
+
+  # temporary fix for iostream issue
+  NIX_CFLAGS_COMPILE = stdenv.lib.optionalString stdenv.isDarwin "-I${libcxx}/include/c++/v1";
```
@copumpkin Can I make everything depending on xcbuild have this flag?

matthewbauer commented on pkgs/os-specific/darwin/apple-source-releases/default.nix:
```diff
@@ -213,6 +218,11 @@ let
     Security        = applePackage "Security"          "osx-10.9.5"      "1nv0dczf67dhk17hscx52izgdcyacgyy12ag0jh6nl5hmfzsn8yy" {};
     xnu             = applePackage "xnu"               "osx-10.11.6"     "0yhziq4dqqcbjpf6vyqn8xhwva2zb525gndkx8cp8alzwp76jnr9" {};
+    Librpcsvc       = applePackage "Librpcsvc"         "osx-10.11.6"     "1zwfwcl9irxl1dlnf2b4v30vdybp0p0r6n6g1pd14zbdci1jcg2k" {};
+    adv_cmds        = applePackage "adv_cmds/xcode.nix" "osx-10.11.6"    "12gbv35i09aij9g90p6b3x2f3ramw43qcb2gjrg8lzkzmwvcyw9q" {};
```
This might be confusing with ps, and locale up top.

copumpkin replied:
Yeah, I think once we merge this we'll need to go reconcile all this stuff 😄

matthewbauer commented:
I'm taking off the work in progress flag for now because this should be ready to go. I think it still has room for improvement but it may be easiest to merge sooner rather than later.

matthewbauer committed:
xcbuild: Move extra specs to platform

copumpkin commented on pkgs/development/tools/xcbuild/default.nix:
```diff
@@ -39,5 +39,6 @@ in stdenv.mkDerivation rec {
-}
\ No newline at end of file
+  buildInputs = [ cmake zlib libxml2 libpng ninja ]
+    ++ stdenv.lib.optionals stdenv.isDarwin [ CoreServices CoreGraphics ImageIO ];
```
Do you actually have this working on linux? 😮 I wouldn't block on it if not, but if so, cool 👍

matthewbauer replied:
Well the "xcbuild.raw" builds fine. You can even make it build the wrapper but it won't be able to produce anything because of some mismatch with the clang flags.

matthewbauer replied:
So for example "basic_cmds" will fail with this:

```
    /nix/store/rfzsjjklrp3kw4l8b06jz8f30ddghxf2-nixpkgs.xctoolchain/usr/bin/clang -x c -arch x86_64 -isysroot /nix/store/f9phbmvnwfsmp797p1vbj7xpi928d0w5-nix.nixpkgs.sdk -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -fasm-blocks -fstrict-aliasing -fpascal-strings -Os -g -Wno-trigraphs -Wno-empty-body -Winfinite-recursion -Wdeprecated-declarations -Wno-missing-field-initializers -Wno-newline-eof -Wno-missing-prototypes -Wpointer-sign -Wno-return-type -Wswitch -Wno-four-char-constants -Wno-conversion -Wno-bool-conversion -Wno-constant-conversion -Wno-enum-conversion -Wno-int-conversion -Wno-sign-conversion -Wno-shorten-64-to-32 -Wno-missing-braces -Wparentheses -Wno-shadow -Wno-uninitialized -Wno-unknown-pragmas -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wunused-value -Wno-unused-variable -I./basic_cmds-cnlhhgwxelhhebfdskrslcepmqdd/Build/Products/Release/include -I./basic_cmds-cnlhhgwxelhhebfdskrslcepmqdd/Build/Intermediates/basic_cmds.build/Release/write.build/DerivedSources/x86_64 -I./basic_cmds-cnlhhgwxelhhebfdskrslcepmqdd/Build/Intermediates/basic_cmds.build/Release/write.build/DerivedSources -F./basic_cmds-cnlhhgwxelhhebfdskrslcepmqdd/Build/Products/Release -MMD -MT dependencies -MF ./basic_cmds-cnlhhgwxelhhebfdskrslcepmqdd/Build/Intermediates/basic_cmds.build/Release/write.build/Objects-normal/x86_64/write.d -c /tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c -o ./basic_cmds-cnlhhgwxelhhebfdskrslcepmqdd/Build/Intermediates/basic_cmds.build/Release/write.build/Objects-normal/x86_64/write.o

clang-3.7: warning: argument unused during compilation: '-arch x86_64'
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:39:1: error: unknown type name '__unused'
__unused static const char copyright[] =
^
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:39:10: error: expected identifier or '('
__unused static const char copyright[] =
         ^
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:50:9: error: expected parameter declarator
__RCSID("$FreeBSD: src/usr.bin/write/write.c,v 1.17 2002/09/04 23:29:09 dwmalone Exp $");
        ^
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:50:9: error: expected ')'
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:50:8: note: to match this '('
__RCSID("$FreeBSD: src/usr.bin/write/write.c,v 1.17 2002/09/04 23:29:09 dwmalone Exp $");
       ^
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:50:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
__RCSID("$FreeBSD: src/usr.bin/write/write.c,v 1.17 2002/09/04 23:29:09 dwmalone Exp $");
^
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:198:10: warning: implicitly declaring library function 'strlcpy' with type 'unsigned long (char *, const char *, unsigned long)'
                        (void)strlcpy(atty, u->ut_line, sizeof(atty));
                              ^
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:198:10: note: include the header <string.h> or explicitly provide a declaration for 'strlcpy'
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:280:8: warning: implicit declaration of function 'time' is invalid in C99 [-Wimplicit-function-declaration]
        now = time((time_t *)NULL);
              ^
/tmp/nix-build-basic_cmds-osx-10.11.6.drv-0/basic_cmds-55/write/write.c:281:9: warning: implicit declaration of function 'ctime' is invalid in C99 [-Wimplicit-function-declaration]
        nows = ctime(&now);
               ^
4 warnings and 4 errors generated.
```

grp replied:
I can explain that! Nothing in that command line says what target to build for, so it's defaulting to the host. And `-arch` is a Darwin-specific flag, and `__unused` is a Darwin extension, hence the errors.

The way xcbuild / Xcode says what platform to build for is through the `-m<platform>-version-min=` flag, which both sets the deployment target and the target triple (usually set by `-target <triple>`). Those are controlled by some options from the platform's `Info.plist`.

Here's what it should be to build for macOS:

    DefaultProperties = {
        DEPLOYMENT_TARGET_SETTING_NAME = "MACOSX_DEPLOYMENT_TARGET";
        DEPLOYMENT_TARGET_CLANG_ENV_NAME = "MACOSX_DEPLOYMENT_TARGET";
        DEPLOYMENT_TARGET_CLANG_FLAG_NAME = "macosx-version-min";
    };

matthewbauer replied:
That's really helpful. Honestly, I'm not sure what I was expecting it to do on Linux. I suppose I thought it might be able to just compile to the host without any modification. I think we always just assume it's compiling to the native environment. Is there any way to force clang to accept macOS args and extensions but actually compile to Linux, I wonder? Not that any one on Linux is dying to use Apple's "write" command or anything but it would be really interesting to see what happens.

We do have some "cross compiling" in Nixpkgs support. I really don't understand how it works enough to set it up, though. It would be really interesting to try to link this into xcbuild.

copumpkin commented on pkgs/development/tools/xcbuild/developer.nix:
```diff
@@ -0,0 +1,18 @@
+{ stdenv, platform, toolchain, xcbuild, writeText }:
+
+stdenv.mkDerivation {
+  name = "Xcode.app";
```
Seems like a bit of an odd name for the derivation! I'm not sure what this is doing, actually. It's just tying together the `xcbuild` specs with the toolchain spec and so on? Given that you're `cp`ing from `xcbuild`, could it just be a separate output on the original `xcbuild` derivation?

matthewbauer replied:
The name can be changed but the main reason this is here is that xcbuild/xcodebuild are set up to have a "root dir" kind of thing that all other paths are based on. This corresponds to `/Applicatinos/Xcode.app` if you have Xcode installed. We can definitely try to do something more clever with this like setting it to $out/Developer in wrapper.nix and linking `${xcbuild.raw.specs}` or something like that. The last time I tried something like this, though, it seemed to break everything. That was a while back so maybe it is still worth a try now.

grp replied:
I'd call it a "developer root" of some kind. "Xcode.app" includes a developer root but also the IDE and such, so might be less clear.

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,170 @@
+      Type = "Architecture";
+      Name = "Standard Architectures (including 64-bit)";
+      RealArchitectures = [ "x86_64" ];
+		  ArchitectureSetting = "ARCHS_STANDARD_INCLUDING_64_BIT";
```
Wonky indentation 😄

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,170 @@
+    }
+  ];
+
+  PackageTypes = [
```
I'm not necessarily opposed to putting this stuff in here, but could you add some comments on this nix-plist stuff explaining whether it differs from upstream in meaningful ways, or in general how customized it is?

I'm thinking a few months down the line when @grp has made improvements upstream and we want to pull more stuff in from `xcbuild` and I'll be unsure whether this stuff is picked from upstream with minor changes, or hand-written from scratch for our purposes, or is literally identical to upstream except in Nix rather than plists, etc.

Does that make sense?

grp replied:
These specifications are part of the platform, so aren't likely to ever show up in xcbuild. Some of the others might, though; will need to read through again.

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,170 @@
+    mkdir -p $out/
+    cd $out/
+
+    plutil -convert xml1 -o Info.plist ${writeText "Info.plist" (builtins.toJSON Info)}
```
If we end up doing a lot of this, I'd probably make a `writePlist` function somewhere that wraps this process (the `plutil . writeText . toJSON` composition), then symlink its output here.

copumpkin commented on pkgs/development/tools/xcbuild/wrapper.nix:
```diff
@@ -0,0 +1,64 @@
+in
+
+stdenv.mkDerivation {
+  name = "xcbuild-wrapper";
```
I wouldn't bother 😄

copumpkin commented on pkgs/os-specific/darwin/apple-source-releases/adv_cmds/xcode.nix:
```diff
@@ -0,0 +1,37 @@
+    done
+
+    mkdir -p $out/System/Library/LaunchDaemons
+    install fingerd/finger.plist $out/System/Library/LaunchDaemons
```
I don't think this will work, right (it has a hardcoded path to /usr)? I wouldn't include it for now.

matthewbauer replied:
Yeah, this was just for completeness because once "xcodebuild install" is used it should just perform this step automatically.

copumpkin commented on pkgs/os-specific/darwin/apple-source-releases/network_cmds/default.nix:
```diff
@@ -0,0 +1,39 @@
+  NIX_CFLAGS_COMPILE = " -I./unbound -I${xnu}/Library/Frameworks/System.framework/Headers/ -D__APPLE_USE_RFC_2292 -DIPV6_DONTFRAG=14 -DINET6";
+
+  patchPhase = ''
+    # disable spray, rtadvd
```
Ideally just like pkill I'd put a little note explaining why these need to be disabled, so someone can know whether to try to enable them in future.

copumpkin commented on pkgs/tools/misc/contacts/default.nix:
```diff
@@ -9,15 +9,15 @@ stdenv.mkDerivation rec {
+    cp ./contacts-*/Build/Products/Default-*/contacts $out/bin
   '';
 
+  NIX_LDFLAGS = " -F${Foundation}/Library/Frameworks/ -F${AddressBook}/Library/Frameworks/";
```
Surprised this is needed, since frameworks get a setup hook that does almost exactly this. It might pay to look into why that isn't happening, or maybe leave a comment saying "FIXME: wtf why isn't the framework setup hook working" 😄

matthewbauer replied:
There's definitely something weird going on. "Foundation" should be in stdenv anyway right?

copumpkin replied:
Sorry, email lost context of thread. I meant that no, Foundation isn't part of stdenv. CoreFoundation is. First of all, very few unix-flavored packages (of the sort we need in the stdenv) use Foundation (more use CoreFoundation, but even that isn't super common), and second of all, it makes the stdenv "unfree" from Hydra's viewpoint, which would be pretty awful 😄

matthewbauer replied:
Could this be caused by messing with buildPhase in xcbuild?

copumpkin reviewed (changes requested):
Generally looks great, thanks! Main thing I'd ask for before merging is a few comments here and there (I ask for specifics in my comments) for other maintainers of this stuff, mostly along the lines of your thought process for a few of these pieces. If you'd rather just merge now and we can deal with that later we can also do that.

copumpkin commented:
Nope definitely not in stdenv. CoreFoundation is in stdenv.
On Thu, Nov 10, 2016 at 15:10 Matthew Justin Bauer <notifications@github.com>
wrote:

> *@matthewbauer* commented on this pull request.
> ------------------------------
>
> In pkgs/tools/misc/contacts/default.nix
> <https://github.com/NixOS/nixpkgs/pull/16540>:
>
> >    '';
>
> +  NIX_LDFLAGS = " -F${Foundation}/Library/Frameworks/ -F${AddressBook}/Library/Frameworks/";
>
> There's definitely something weird going on. "Foundation" should be in
> stdenv anyway right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/16540>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKP3aa7C3kREX_1Qd9ky-PR3qWYVQjks5q83pPgaJpZM4I-sxO>
> .
>

matthewbauer committed:
xcbuild: fix toolchain on Linux

matthewbauer committed:
xcbuild: update meta and comments

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,170 @@
+      ExecDescription = "Yacc $(InputFile)";
+    }
+    {
+      Identifier = "com.apple.compilers.lex";
```
I think it generally looks good now, but did you get my comment earlier asking for some small commentary on where these things come from? Something like "This based mostly on https://github.com/facebook/xcbuild/tree/master/Specifications blah blah blah, but I had to change X for reasons Y and Z"

matthewbauer commented:
So, I'm just going to wait or facebook/xcbuild#183 to be merged in. I'll update the hashes as soon as it is.

matthewbauer committed:
xcbuild: Get rid of developer.nix, move to wrapper

also:

- add custom outputs "specs" for xcbuild
- get rid of unneeded tools
- update xcbuild
- add more comments
- fixup xcbuild derivations

Affected xcbuild derivations include:

- adv_cmds
- network_cmds
- basic_cmds

copumpkin commented on pkgs/development/tools/xcbuild/platform.nix:
```diff
@@ -0,0 +1,152 @@
+  # /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Xcode/Speciications/.
+
+  # Based off of the MacOSX Architectures.xcpsec file. All i386 stuff
+  # is removed because NixPkgs only supports darwin-x86_64.
```
👍

copumpkin commented:
@matthewbauer I pinged you on IRC, but if you didn't notice, @grp merged our favorite PR into xcbuild 👍

matthewbauer commented:
@copumpkin Yep, this PR now reflects that and should be ready to merge.

copumpkin commented:
Also not sure if you saw my comments on IRC, but I think I know what's wrong with adv_cmds and the various c++ issues we've been seeing. Basically our cc-wrapper looks for information about where to find libc++ from an environment variable, and `xcbuild` doesn't pass that variable (or related ones) through. I haven't had time to tinker with it, but I think that'll resolve a few issues we've been seeing.

matthewbauer commented:
@copumpkin So it should pass "NIX_CFLAGS_COMPILE" and "NIX_LDFLAGS" in OTHER_FLAGS, I think.

copumpkin commented:
The ones that I think matter in this case are https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L114-L116

Basically, `NIX_CXXSTDLIB_COMPILE`, `NIX_CXXSTDLIB_LINK`, and perhaps making the default `default_cxx_stdlib_compile` into something non-empty on darwin. I hate how complicated the wrapper is 😦. cc @shlevy

copumpkin commented:
For reference: https://github.com/facebook/xcbuild/issues/194
